src/core/reference/include/openvino/reference/fake_quantize.hpp

```diff
@@ -55,133 +110,451 @@ void fake_quantize(const T* const arg,
                    const op::AutoBroadcastSpec& broadcast) {
     using namespace fake_quantize_details;
 
+    T levels_minus_one = static_cast<T>(levels - 1);
+    const size_t arg_size = shape_size(arg_shape);
+
     if (shape_size(in_low_shape) == 1 && shape_size(in_high_shape) == 1 && shape_size(out_low_shape) == 1 &&
         shape_size(out_high_shape) == 1) {
-        const size_t arg_size = shape_size(arg_shape);
-        const auto q = [=](const T& a) {
-            return quantize(a, *in_low, *in_high, *out_low, *out_high, levels);
-        };
         for (size_t i = 0; i < arg_size; ++i) {
-            out[i] = q(arg[i]);
-        }
-    } else {
-        OPENVINO_ASSERT(in_low_shape.size() <= arg_shape.size() && in_high_shape.size() <= arg_shape.size() &&
-                            out_low_shape.size() <= arg_shape.size() && out_high_shape.size() <= arg_shape.size(),
-                        "Tensors with input\\output ranges should have rank less or "
-                        "equal to data tensor rank equal to ",
-                        arg_shape.size());
-
-        Shape arg0_padded_shape = arg_shape;
-        Shape arg1_padded_shape = in_low_shape;
-        Shape arg2_padded_shape = in_high_shape;
-        Shape arg3_padded_shape = out_low_shape;
-        Shape arg4_padded_shape = out_high_shape;
-
-        size_t max_shape_size = arg_shape.size();
-
-        while (arg0_padded_shape.size() < max_shape_size) {
-            arg0_padded_shape.insert(arg0_padded_shape.begin(), 1);
+            out[i] = quantize(arg[i], *in_low, *in_high, *out_low, *out_high, levels_minus_one);
         }
+        return;
+    }
 
-        while (arg1_padded_shape.size() < max_shape_size) {
-            arg1_padded_shape.insert(arg1_padded_shape.begin(), 1);
-        }
+    // clang-format off
+    /*
+     * ---------------------------------------------------
+     * Overview:
+     *  Numpy broadcasted input tensors can be partitioned into two: outer and inner part (which also defines inner
+     *  stride as a product of inner part), so N-dimensional tensors can be processed using two loops.
+     *
+     *  For example with two inputs [2, 2, 3, 4] and [1, 1, 3, 4] we can have:
+     *      input 1 with shape [2, 2, 3, 4] can be divided into outer part [2, 2] and inner part [4, 5]
+     *      with inner stride = 12 (3 * 4) input 2 with shape [1, 1, 3, 4] can be divided into outer part [1, 1]
+     *      and inner part [4, 5] with inner stride = 12 (3 * 4)
```

Comment:
Maybe I missed something but looks like inner part should be `[3, 4]`
```suggestion
     *      input 1 with shape [2, 2, 3, 4] can be divided into outer part [2, 2] and inner part [3, 4]
     *      with inner stride = 12 (3 * 4) input 2 with shape [1, 1, 3, 4] can be divided into outer part [1, 1]
     *      and inner part [3, 4] with inner stride = 12 (3 * 4)
```

Reply:
you're right. I've had different values initially - looks like I failed to change that part of the comment.